extension/src/cli/git/reader.ts

```diff
@@ -118,8 +118,20 @@ export class GitReader extends GitCli {
       env: { LANG: 'en_US.UTF-8' }
     })
     try {
-      const branches = await this.executeProcess(options)
-      return trimAndSplit(branches)
+      const branchesOutput = await this.executeProcess(options)
+      const branches = trimAndSplit(branchesOutput)
+      const parsedBranches = []
+      const noBranchText = '(no branch)'
+
+      for (const branch of branches) {
+        if (branch.includes(noBranchText)) {
+          const sha = await this.revParseHead(cwd)
+          parsedBranches.push(branch.replace(noBranchText, sha))
```

Comment:
[C] We now have two sets of logic. One here and one at: https://github.com/iterative/vscode-dvc/blob/a2920aa84f4f25a6b91bb16b6d111ac8f4dbb5ba/extension/src/experiments/model/collect.ts#L497-L501

Maybe that is ok but it would be good to pull out constants like `HEAD` & `(no branch)` as they may serve as breadcrumbs later. We could also move `isCurrentBranch` into the `git` folder.

A good practice would be to (generally) encapsulate all of the logic in the same place. I've tried to leave the reader as dumb as possible previously and not put any transformation logic in here at all (an exception being trim and split). The reason for encapsulation is that in future when things inevitably break we won't be searching through the code to work out all of the possible places something has been changed.

With all of that said maybe the correct place for this logic is in:

https://github.com/iterative/vscode-dvc/blob/a2920aa84f4f25a6b91bb16b6d111ac8f4dbb5ba/extension/src/experiments/data/index.ts#L125-L132

after calling `setBranches` check the current branch and overwrite with `revParseHead` if it matches `(no branch)` this is a special case so we should still avoid calling `revParseHead` if necessary. 

Does all of that make sense?

Reply:
Another (more involved) way of fixing the issue is assuming that these commands:

https://github.com/iterative/vscode-dvc/blob/a2920aa84f4f25a6b91bb16b6d111ac8f4dbb5ba/extension/src/experiments/data/index.ts#L85-L97

will work if we don't pass a branch to them (needs tested). We could then leave the detached HEAD/no branch strings as is (I think) and not pass a branch for the current branch. If this all worked as expected it should lead to a slightly better UX of showing the original branch name in the experiments table (e.g `(no branch)` so that they user knows that they currently don't have a branch to work with.

Reply:
> Does all of that make sense?

Yes, makes sense! I'll move this branch parsing logic to `updateBranches`.

> will work if we don't pass a branch to them (needs tested). We could then leave the detached HEAD/no branch strings as is (I think) and not pass a branch for the current branch. If

Did a quick test and `GIT_GET_NUM_COMMITS` needs a branch to work with.

Reply:
`GIT_GET_NUM_COMMITS` <= could this use `HEAD`?

Reply:
Making these updates seems to work (rough): 

```
  private async updateExpShow() {
    await this.updateBranches()
    const [currentBranch, ...branches] = this.experiments.getBranchesToShow()
    const availableNbCommits: { [branch: string]: number } = {}

    const promises = []

    promises.push(
      this.collectGitLogByBranch(currentBranch, availableNbCommits, true)
    )

    for (const branch of branches) {
      promises.push(this.collectGitLogByBranch(branch, availableNbCommits))
    }

    const branchLogs = await Promise.all(promises)
    const { args, gitLog, rowOrder } = this.collectGitLogAndOrder(branchLogs)

    const expShow = await this.internalCommands.executeCommand<ExpShowOutput>(
      AvailableCommands.EXP_SHOW,
      this.dvcRoot,
      ...args
    )

    this.collectFiles({ expShow })
    return { availableNbCommits, expShow, gitLog, rowOrder }
  }

  private async collectGitLogByBranch(
    branch: string,
    availableNbCommits: { [branch: string]: number },
    isCurrent?: boolean
  ) {
    const nbOfCommitsToShow = this.experiments.getNbOfCommitsToShow(branch)

    const [branchLog, totalCommits] = await Promise.all([
      this.internalCommands.executeCommand(
        AvailableCommands.GIT_GET_COMMIT_MESSAGES,
        this.dvcRoot,
        isCurrent ? 'HEAD' : branch,
        String(nbOfCommitsToShow)
      ),
      this.internalCommands.executeCommand<number>(
        AvailableCommands.GIT_GET_NUM_COMMITS,
        this.dvcRoot,
        isCurrent ? 'HEAD' : branch
      )
    ])

    availableNbCommits[branch] = totalCommits

    return {
      branch,
      branchLog: isCurrent ? branchLog.replace('HEAD', branch) : branchLog
    }
  }

...

  private async updateBranches() {
    const allBranches = await this.internalCommands.executeCommand<string[]>(
      AvailableCommands.GIT_GET_BRANCHES,
      this.dvcRoot
    )

    const { currentBranch, branches } = collectBranches(allBranches)

    this.experiments.setBranches(branches, currentBranch)
  }


```

<img width="1728" alt="image" src="https://github.com/iterative/vscode-dvc/assets/37993418/dc478cf1-dc9a-44b7-9226-25efb7ce6a22">


Happy to pick up this second change if I won't be stepping on your toes.

Reply:
note: also need to update `collectBranches`:

```
    const cleanBranch = branch.replace('* ', '')
    // .replace(new RegExp(`\\(${gitPath.DOT_GIT_HEAD}\\s\\w+\\s\\w+\\s`), '')
    // .replace(')', '')

```

Reply:
Ohhhhhh, ok, I understand your solution a lot better, thanks! I'll apply your changes + add tests.